### DISCUSSION TITLE
Error with slightly modified Simple_Patch_Antenna script

### DISCUSSION THREAD

**Discussion by DatKnight** [thliebig/openEMS-Project#209] [General]
I've had a go at trying to simulate a Patch Antenna in openEMS, starting by just modifying the existing Simple_Patch_Antenna.m script. The only changes I've made are to the patch and substrate dimensions and dielectric constant:

![image](https://github.com/thliebig/openEMS-Project/assets/25912106/251c4dc1-e667-4b3d-adab-657d15044631)

But when I run the script I get the following error:

![image](https://github.com/thliebig/openEMS-Project/assets/25912106/3cae5421-0da4-43fc-8458-da32ddf3735c)

Thing is if I inspect the port variable:
![image](https://github.com/thliebig/openEMS-Project/assets/25912106/a9c6d41f-711e-4ec9-9b53-7b7f95241c4e)

So I've got no clue why it throws an error as I've not messed around with the port dimensions, and the port should still be on the PCB?

**Comment by biergaizi** [comment]
> The only changes I've made are to the patch and substrate dimensions

This is the red flag. Many simulation setups can be surprisingly sensitive to dimensions, and there are many traps for new users: wrong meshing, wrong port locations, ports or components misaligned to meshes, coordinates too close to absorbing boundary conditions, violation of the 1/3-2/3 rule...

In this case, the number of cells on the Z axis is probably too low, just barely enough to finish the original simulation:

    substrate.cells = 4;

Once the substrate height is slightly increased, the mesh became unable to correctly cover the port on the Z axis, so the port stops working.

 To fix that, use:

    substrate.cells = 10;

**Comment by DatKnight** [comment]
That worked! Thank you so much!

Is there anywhere I can find more about these rules of thumb/traps/quirks with the software? There isn't much I can find on the wiki that is in that level of detail, am I mostly just limited to trial and error and asking questions on here?

**Reply by biergaizi** [reply]
Unfortunately, no. Speaking as an unofficial project contributor, I think we should eventually write a new FAQ and tutorial that covers these aspects. As for now,  feel free to post on the forums if you have any new question.

Here are some useful links in the forum about coordinates and meshing related question:

* Correct Usages of Boundary Conditions: https://github.com/thliebig/openEMS-Project/discussions/159
* Optimal microstrip lumped port size/placement and Thirds Rule question: https://github.com/thliebig/openEMS-Project/discussions/200
* Weird Behavior of Lumped Components: https://github.com/thliebig/openEMS-Project/discussions/186
* Lumped port: partial overlap with conductor metal allowed?: https://github.com/thliebig/openEMS-Project/discussions/162

Also, be sure to check the archive of the now-closed old project Wiki - many documents are not rewritten for the new website yet...

* https://wiki.openems.de/index.php/Main_Page.html

**Reply by thliebig** [reply]
Yeah, having a full list of all this necessary knowledge would be nice, but I guess it will be never complete. But I want to point out that it not as much about "this software" as for the FDTD method and EM-modeling ;)

**Reply by DatKnight** [reply]
Are there any good resources for learning more about that?
I've learned some at uni but never went too in detail about the specifics of EM modelling or the FDTD method

**Reply by biergaizi** [reply]
https://docs.openems.de/more.html

**Reply by LubomirJagos42** [reply]
maybe this is not best advice but since I start playing with openEMS I had to walk through all that caveats and so and most frustrating is building up simulation from scripting, like but my opinion to make simulation build up from ground without knowing API is putting there some layer of unknown and after that there are other layers of unknow so I created GUI for FreeCAD to generate these octave or python scripts similar way how HFSS used to define simulation and put there lot of user input checks to generate at least runnable simulation you can look here:
    https://github.com/LubomirJagos42/FreeCAD-OpenEMS-Export
here are tested examples:
    https://github.com/LubomirJagos42/FreeCAD-OpenEMS-Export-Plugin-Documentation
    
if you want to walk through this trial and error way faster in words to put more effort into building it visualy and at least to have clue how I solved some caveats for simulations this will help you got oriented in openEMS faster

definitely what you should do is ALWAYS check your simulation in AppCSXCAD and think what it's doing, this is just to provide some way from top to bottom like some guide how you can implement your simulation, let say inspiration

when you master scripting way you can always do your model in scripting, hope it helps

**Reply by biergaizi** [reply]
> ALWAYS check your simulation in AppCSXCAD and think what it's doing

I'd like to add that it's also often a good idea to enable field dump so one can visually examine what the simulation is doing in ParaView as a 3D animation - which is extremely valuable for debugging malfunctioning simulations.

**Comment by DatKnight** [comment]
I've changed the simulation again to move the feed point to the edge of the patch and I'm getting the same error :l

Been having a look at trying to see what might be wrong, but for some reason the port isn't showing up in AppCSXCAD?
![image](https://github.com/thliebig/openEMS-Project/assets/25912106/047ebe85-ae99-416e-b2aa-cfdd0b4b2550)

It should be turned on.

There's also some errors that pop up in the console when I try to rotate the view - I've been making do by just looking at the xz/xy/zy planes as it lags it the hell out when I rotate with the mouse.

![image](https://github.com/thliebig/openEMS-Project/assets/25912106/6ddb8254-5621-421b-b5db-15ef4fb22a29)

I've googled it and it seems to be a generic issue, I'm not sure if it's something anyone here as encountered before?

@biergaizi I've tried having a look at adding a dump box for Et fields and viewing those:

![image](https://github.com/thliebig/openEMS-Project/assets/25912106/e695354f-7640-4631-a323-2055e3ae1b8a)

But I just get an empty box with nothing showing up when I select E-Field and run the animation, this also happens with the original script with unmodified parameters, even though the simulation has no problems.

![image](https://github.com/thliebig/openEMS-Project/assets/25912106/e752a690-7f0c-4fec-9d77-0ea20920acd0)

I'm really confused here, will going and doing my homework to understand the maths help me use the software and stop so much of this trial and error? I feel like I should be able to get something simple like a Patch antenna working without too much hassle.

**Reply by biergaizi** [reply]
> but for some reason the port isn't showing up in AppCSXCAD?

The port is not visible because you didn't highlight it. You should click the "lightbulb" symbol on all port-related structures to enable them. Even then, the port is only a few pixels large within the 3D model so you need to zoom in a couple of times to see it.

> I've been making do by just looking at the xz/xy/zy planes as it lags it the hell out when I rotate with the mouse. I've googled it and it seems to be a generic issue, I'm not sure if it's something anyone here as encountered before?

I don't think anyone has encountered that before. It says 3D acceleration is not working on your system, OpenGL GLEW cannot be initialized due to an unknown error. Thus, it falls back to software rendering on the CPU, which would run at a speed of a few FPS at best. This problem is related to your machine and operating system. Without knowledge about your specific system environment, there's nothing we can do. Please tell us about your system - Which distribution is it? Is it virtualized or bare-metal? Does it has an iGPU or a dedicated GPU, are any other 3D application functioning correctly in your system.

Also, note that this problem has nothing to do with openEMS, I'm only trying on answer your questions on a best-effort basis.

> But I just get an empty box with nothing showing up when I select E-Field and run the animation

Nothing was showing up because you didn't select anything. You need to click the *Apply* button on the left side of the screen in order to apply the dataset `E-Field` into the current visualization. After clicking *Apply*, you then need to select a suitable data item using the *Coloring* option. By default, it's "Solid Color", so nothing would show up. You need to click the menu and change it to `E-Field` (or other kind of dumps, depending on the simulation setup). You can also change the type of representation using a drop-down menu at the top of the screen, there are *Surface*, *Volume*, *Wireframe*, etc. 

Again, this problem has nothing to do with openEMS. It a ParaView question, which is the standard tool in the scientific computing world, basic knowledge is mandatory for everyone working in the field of scientific data visualization. 

> I'm really confused here, will going and doing my homework to understand the maths help me use the software and stop so much of this trial and error? I feel like I should be able to get something simple like a Patch antenna working without too much hassle.

As you can see, there many problems involved, both theoretical and practical - math problems, familiarity of common tools like ParaView, etc. The fact that Qt's OpenGL is malfunctioning on your system certainly doesn't help.

At the end of the day, openEMS is only a simulation engine primarily developed as a pure research project, it's *not* a full-featured EDA system, it's also not a full-featured antenna design CAD.  The engine has no idea about what you're simulating, it just accepts an arbitrary 3D model with material data, then applies Maxwell's equations (this is the simplicity of the FDTD method, which is its great advantage). Given the correct inputs, it can in principle model everything. 

But there is no PCB CAD, no visualization tools, no pre-supplied 3D model, no auto-meshing. In each simulation, you're essentially recreating the universe from scratch, one polygon at a time. Thus, I recommend new users not to be overambitious. Also, after a simulation is complete, there's no built-in post-processing capabilities (besides near-field to far-field transformation), you get raw data and it's up to your to analyze and interpret the results. For myself, I use the Python library `scikit-rf`, which is quite a powerful tool to analyze S-parameter data in the frequency domain (But you're using Octave/Matlab and I don't know what kind of tools are available). Thus, there are no such things as "simple tasks". 

Personally, I spent 6 months on and off on my spare time before I learned the basis. I think simulating an antenna is already a difficult starting point of using openEMS. It's better to model something even simpler in order to get familiar with the basic concept first - such as a modeling a single microstrip transmission line on top of a ground plane.

I acknowledge that the documentation is severely lacking at the moment. I'd like to write a practical guide to simulation from the perspective of a circuit designer, but I currently don't have the time.

**Reply by LubomirJagos42** [reply]
@DatKnight I think that error what you are getting is due you redefine something and cube is reduced to point, also in Paraview you have just E-field at step 0 so you get nothing drawn :(

Can you please attach here your whole script? It will be much easier to examine it as whole as just trying to guess where is problem.

openEMS examples are all done good and getting right results, but what is little confusing on them is when you just start with openEMS that they are parametric models and if you change something it also change grid and so so there is too many things which are modified after, also feedpoints are mostli attached to just one gridline and almost invisible in 3D view but they are there you just need to zoom to them :D :D

just put here your script and I bet that there is just some simple problem which can be solved fast if whole simulation available

**Reply by DatKnight** [reply]
> The port is not visible because you didn't highlight it. You should click the "lightbulb" symbol on all port-related structures to enable them. Even then, the port is only a few pixels large within the 3D model so you need to zoom in a couple of times to see it.

It still doesn't show up when I enable the lightbulb, I've figured out that its because the port in the original script is 1D, it shows up when I add some width to it:
 
![image](https://github.com/thliebig/openEMS-Project/assets/25912106/aad6652c-bec7-4cf2-8a2d-54a9e909f4a2)

> I don't think anyone has encountered that before. It says 3D acceleration is not working on your system, OpenGL GLEW cannot be initialized due to an unknown error. Thus, it falls back to software rendering on the CPU, which would run at a speed of a few FPS at best. This problem is related to your machine and operating system. Without knowledge about your specific system environment, there's nothing we can do. Please tell us about your system - Which distribution is it? Is it virtualized or bare-metal? Does it has an iGPU or a dedicated GPU, are any other 3D application functioning correctly in your system.

My specs are:
![image](https://github.com/thliebig/openEMS-Project/assets/25912106/40fc76eb-6afb-4a90-96f6-506de4600f38)

My CPU isn't that out of date, I imagine a 7 year old CPU presumably shouldn't have any issues with OpenGL.


> Nothing was showing up because you didn't select anything. You need to click the _Apply_ button on the left side of the screen in order to apply the dataset `E-Field` into the current visualization. After clicking _Apply_, you then need to select a suitable data item using the _Coloring_ option. By default, it's "Solid Color", so nothing would show up. You need to click the menu and change it to `E-Field` (or other kind of dumps, depending on the simulation setup). You can also change the type of representation using a drop-down menu at the top of the screen, there are _Surface_, _Volume_, _Wireframe_, etc.
> 
> Again, this problem has nothing to do with openEMS. It a ParaView question, which is the standard tool in the scientific computing world, basic knowledge is mandatory for everyone working in the field of scientific data visualization.
>

I'll spend some more time trying to understand Paraview then.

> Personally, I spent 6 months on and off on my spare time before I learned the basis. I think simulating an antenna is already a difficult starting point of using openEMS. It's better to model something even simpler in order to get familiar with the basic concept first - such as a modeling a single microstrip transmission line on top of a ground plane.

That makes me feel better, maybe I'll try going back to my cavity simulation for now, at least it's not just me.
 
> I acknowledge that the documentation is severely lacking at the moment. I'd like to write a practical guide to simulation from the perspective of a circuit designer, but I currently don't have the time.

Maybe once/if I get a better grasp on things I'll have a go at writing some documentation myself, would be nice to contribute.

**Reply by DatKnight** [reply]
@LubomirJagos42 here is my code

```
close all
clear
clc

%% Setup the Simulation
physical_constants;
unit = 1e-3; % all length in mm

% patch width in x-direction
patch.width  = 38; % resonant length
% patch length in y-direction
patch.length = 29;

%substrate setup
substrate.epsR   = 4.5;
substrate.kappa  = 1e-3 * 2*pi*2.45e9 * EPS0*substrate.epsR;
substrate.width  = 100;
substrate.length = 100;
substrate.thickness = 1.6;
substrate.cells = 10;

%setup feeding
feed.pos = -14; %feeding position in y-direction
feed.R = 50;     %feed resistance

% size of the simulation box
SimBox = [200 200 150];

%% Setup FDTD Parameter & Excitation Function
f0 = 2e9; % center frequency
fc = 1e9; % 20 dB corner frequency
FDTD = InitFDTD( 'NrTs', 30000 );
FDTD = SetGaussExcite( FDTD, f0, fc );
BC = {'MUR' 'MUR' 'MUR' 'MUR' 'MUR' 'MUR'}; % boundary conditions
FDTD = SetBoundaryCond( FDTD, BC );

%% Setup CSXCAD Geometry & Mesh
CSX = InitCSX();

%initialize the mesh with the "air-box" dimensions
mesh.x = [-SimBox(1)/2 SimBox(1)/2];
mesh.y = [-SimBox(2)/2 SimBox(2)/2];
mesh.z = [-SimBox(3)/3 SimBox(3)*2/3];

% Create Patch
CSX = AddMetal( CSX, 'patch' ); % create a perfect electric conductor (PEC)
start = [-patch.width/2 -patch.length/2 substrate.thickness];
stop  = [ patch.width/2  patch.length/2 substrate.thickness];
CSX = AddBox(CSX,'patch',10,start,stop); % add a box-primitive to the metal property 'patch'

% Create Substrate
CSX = AddMaterial( CSX, 'substrate' );
CSX = SetMaterialProperty( CSX, 'substrate', 'Epsilon', substrate.epsR, 'Kappa', substrate.kappa );
start = [-substrate.width/2 -substrate.length/2 0];
stop  = [ substrate.width/2  substrate.length/2 substrate.thickness];
CSX = AddBox( CSX, 'substrate', 0, start, stop );

% add extra cells to discretize the substrate thickness
mesh.z = [linspace(0,substrate.thickness,substrate.cells+1) mesh.z];

% Create Ground same size as substrate
CSX = AddMetal( CSX, 'gnd' ); % create a perfect electric conductor (PEC)
start(3)=0;
stop(3) =0;
CSX = AddBox(CSX,'gnd',10,start,stop);

% Apply the Excitation & Resist as a Current Source
start = [-1 feed.pos  0];
stop  = [1 feed.pos  substrate.thickness];
[CSX port] = AddLumpedPort(CSX, 5 ,1 ,feed.R, start, stop, [0 0 1], true);

% Finalize the Mesh
% detect all edges except of the patch
mesh = DetectEdges(CSX, mesh,'ExcludeProperty','patch');
% detect and set a special 2D metal edge mesh for the patch
mesh = DetectEdges(CSX, mesh,'SetProperty','patch','2D_Metal_Edge_Res', c0/(f0+fc)/unit/50);
% generate a smooth mesh with max. cell size: lambda_min / 20
mesh = SmoothMesh(mesh, c0/(f0+fc)/unit/20);
CSX = DefineRectGrid(CSX, unit, mesh);

CSX = AddDump(CSX,'Hf', 'DumpType', 11, 'Frequency',[2.4e9]);
CSX = AddBox(CSX,'Hf',10,[-substrate.width -substrate.length -10*substrate.thickness],[substrate.width +substrate.length 10*substrate.thickness]); %assign box

CSX = AddDump(CSX,'Et');
CSX = AddBox(CSX,'Et', 0,[-substrate.width -substrate.length -10*substrate.thickness],[substrate.width +substrate.length 10*substrate.thickness]); %assign box

% add a nf2ff calc box; size is 3 cells away from MUR boundary condition
start = [mesh.x(4)     mesh.y(4)     mesh.z(4)];
stop  = [mesh.x(end-3) mesh.y(end-3) mesh.z(end-3)];
[CSX nf2ff] = CreateNF2FFBox(CSX, 'nf2ff', start, stop);

%% Prepare and Run Simulation
Sim_Path = 'tmp_Patch_Ant';
Sim_CSX = 'patch_ant.xml';

% create an empty working directory
[status, message, messageid] = rmdir( Sim_Path, 's' ); % clear previous directory
[status, message, messageid] = mkdir( Sim_Path ); % create empty simulation folder

% write openEMS compatible xml-file
WriteOpenEMS( [Sim_Path '/' Sim_CSX], FDTD, CSX );

% show the structure
CSXGeomPlot( [Sim_Path '/' Sim_CSX] );

% run openEMS
RunOpenEMS( Sim_Path, Sim_CSX);

%% Postprocessing & Plots
freq = linspace( max([1e9,f0-fc]), f0+fc, 501 );
port = calcPort(port, Sim_Path, freq);

%% Smith chart port reflection
plotRefl(port, 'threshold', -10)
title( 'reflection coefficient' );

% plot feed point impedance
Zin = port.uf.tot ./ port.if.tot;
figure
plot( freq/1e6, real(Zin), 'k-', 'Linewidth', 2 );
hold on
grid on
plot( freq/1e6, imag(Zin), 'r--', 'Linewidth', 2 );
title( 'feed point impedance' );
xlabel( 'frequency f / MHz' );
ylabel( 'impedance Z_{in} / Ohm' );
legend( 'real', 'imag' );

% plot reflection coefficient S11
s11 = port.uf.ref ./ port.uf.inc;
figure
plot( freq/1e6, 20*log10(abs(s11)), 'k-', 'Linewidth', 2 );
grid on
title( 'reflection coefficient S_{11}' );
xlabel( 'frequency f / MHz' );
ylabel( 'reflection coefficient |S_{11}|' );

drawnow

%% NFFF Plots
%find resonance frequency from s11
f_res_ind = find(s11==min(s11));
f_res = freq(f_res_ind);

% calculate the far field at phi=0 degrees and at phi=90 degrees
disp( 'calculating far field at phi=[0 90] deg...' );

nf2ff = CalcNF2FF(nf2ff, Sim_Path, f_res, [-180:2:180]*pi/180, [0 90]*pi/180);

% display power and directivity
disp( ['radiated power: Prad = ' num2str(nf2ff.Prad) ' Watt']);
disp( ['directivity: Dmax = ' num2str(nf2ff.Dmax) ' (' num2str(10*log10(nf2ff.Dmax)) ' dBi)'] );
disp( ['efficiency: nu_rad = ' num2str(100*nf2ff.Prad./port.P_inc(f_res_ind)) ' %']);

% normalized directivity as polar plot
figure
polarFF(nf2ff,'xaxis','theta','param',[1 2],'normalize',1)

% log-scale directivity plot
figure
plotFFdB(nf2ff,'xaxis','theta','param',[1 2])
% conventional plot approach
% plot( nf2ff.theta*180/pi, 20*log10(nf2ff.E_norm{1}/max(nf2ff.E_norm{1}(:)))+10*log10(nf2ff.Dmax));

drawnow

% Show 3D pattern
disp( 'calculating 3D far field pattern and dumping to vtk (use Paraview to visualize)...' );
thetaRange = (0:2:180);
phiRange = (0:2:360) - 180;
nf2ff = CalcNF2FF(nf2ff, Sim_Path, f_res, thetaRange*pi/180, phiRange*pi/180,'Verbose',1,'Outfile','3D_Pattern.h5');

figure
plotFF3D(nf2ff,'logscale',-20);


E_far_normalized = nf2ff.E_norm{1} / max(nf2ff.E_norm{1}(:)) * nf2ff.Dmax;
DumpFF2VTK([Sim_Path '/3D_Pattern.vtk'],E_far_normalized,thetaRange,phiRange,'scale',1e-3);
```

**Reply by LubomirJagos42** [reply]
ok, thanks, I corrected your simulation:
  - commented smith chart reflection due there is some error and because of that it's not continuing so to get you other graph I disabled that
  - added post_processing at beginning if set to 1 it's not doing simulation, skipping it and continue to graphs
  - E, H probes made planar, not 3D, I set their thickness to 0, also I still don't know how to get slices in Paraview from 3D, but you want planar E, H probe, at least this way you got nice animation in paraview when set as @biergaizi mentioned in previous comment

your design is on good way, S11 is bad, but it's showing there is some resonance :D
![image](https://github.com/thliebig/openEMS-Project/assets/147299357/33ad45e0-513a-4004-acf9-d04877034d4f)

[patch_antenna_DatKnight.txt](https://github.com/thliebig/openEMS-Project/files/15491855/patch_antenna_DatKnight.txt)

**Reply by LubomirJagos42** [reply]
here is animation of E field :D

https://github.com/thliebig/openEMS-Project/assets/147299357/396034bd-0364-4a2f-912f-1c694deb8d1e

**Reply by biergaizi** [reply]
> My CPU isn't that out of date, I imagine a 7 year old CPU presumably shouldn't have any issues with OpenGL.

Thanks for posting detailed spec. I believe the root cause is: https://github.com/nigels-com/glew/issues/172 In short, OpenGL GLEW library is currently incompatible with Wayland because it wants to use X11's GLX, but is unable to find it. So far there's no fix in the stable release. As you can see under that issue, almost everything that uses GLEW are affected, with links to ~20 projects.

For now, try running AppCSXCAD under XWayland instead.

    env WAYLAND_DISPLAY= /path/to/application/you/want/to/run

This forces `WAYLAND_DISPLAY` to be an empty variable.

**Reply by DatKnight** [reply]
> > My CPU isn't that out of date, I imagine a 7 year old CPU presumably shouldn't have any issues with OpenGL.
> 
> Thanks for posting detailed spec. I believe the root cause is: [nigels-com/glew#172](https://github.com/nigels-com/glew/issues/172) In short, OpenGL GLEW library is currently incompatible with Wayland because it wants to use X11's GLX, but is unable to find it. So far there's no fix in the stable release. As you can see under that issue, almost everything that uses GLEW are affected, with links to ~20 projects.
> 
> For now, try running AppCSXCAD under XWayland instead.
> 
> ```
> env WAYLAND_DISPLAY= /path/to/application/you/want/to/run
> ```
> 
> This forces `WAYLAND_DISPLAY` to be an empty variable.

That works great thanks! How did you go about finding the solution may I ask?

**Reply by DatKnight** [reply]
> ok, thanks, I corrected your simulation:
> 
>     * commented smith chart reflection due there is some error and because of that it's not continuing so to get you other graph I disabled that
> 
>     * added post_processing at beginning if set to 1 it's not doing simulation, skipping it and continue to graphs
> 
>     * E, H probes made planar, not 3D, I set their thickness to 0, also I still don't know how to get slices in Paraview from 3D, but you want planar E, H probe, at least this way you got nice animation in paraview when set as @biergaizi mentioned in previous comment
> 
> 
> your design is on good way, S11 is bad, but it's showing there is some resonance :D ![image](https://private-user-images.githubusercontent.com/147299357/334996445-33ad45e0-513a-4004-acf9-d04877034d4f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwODMyMDIsIm5iZiI6MTcxNzA4MjkwMiwicGF0aCI6Ii8xNDcyOTkzNTcvMzM0OTk2NDQ1LTMzYWQ0NWUwLTUxM2EtNDAwNC1hY2Y5LWQwNDg3NzAzNGQ0Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxNTI4MjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MDI4YTA0YTU3YjEyMTgwODU5MTNjYjU1MTg3MWYyMWViZDVhYWU5MTY3ZjRjZTNiNzFlN2UxNjVlYTg3MzU5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.B33m3TgWvQBblKDcBea3elhsS9VI_NR0bDAyWyFPUQ0)
> 
> [patch_antenna_DatKnight.txt](https://github.com/thliebig/openEMS-Project/files/15491855/patch_antenna_DatKnight.txt)

Thanks so much! Never occurred to me to just turn off the Smith Chart first and see if that works lol, I'm still curious why it doesn't work though.
I've got to figure out how to add some matching to the simulation now, hopefully that doesn't cause more issues.
I'll give your FreeCAD addon a try, I want to be able to get to a point where I know what I'm doing scripting wise, but if this eases the learning curve then that's great!

Thanks so much everyone!

**Reply by biergaizi** [reply]
> @LubomirJagos42: E, H probes made planar, not 3D, I set their thickness to 0, also I still don't know how to get slices in Paraview from 3D, 

To get a 2D slice from a 3D model, first open and visualize the model as usual. Then click *Filter* menu, select *Search*. Type *Slice*. Click the matched *Slice* option. In the menu bar at the left, click Z normal, and click Apply. The result is a slice across the Z dimension. To change the location of the slice, change the 3rd coordinate of the *Origin* property. Don't forget to click *Apply*. The only problem I've found is that the dimension is in SI (I think) so the coordinate values are often extremely small and difficult to type - if it's messed up, click *Reset* to undo the unapplied change.

See ParaView documentation for details: https://docs.paraview.org/en/latest/UsersGuid/filteringData.html#slice

![Screenshot_2024-06-01_08-53-41](https://github.com/thliebig/openEMS-Project/assets/1310693/06b13a0e-39fb-4995-baff-aa1a6595ea4b)

> That works great thanks! How did you go about finding the solution may I ask?

Using search engines and making educated guesses from experience.  I just searched the error message on the Web, and clicked several links. One link brought me to the GLEW bug tracker, the reporter mentioned Wayland, so I became suspicious and made an educated guess that it's probably the culprit. Then, I asked your system specs. You said Fedora, so my suspicious was almost confirmed. I then installed a fresh Fedora on a virtual machine and reproduced the problem. I've now submitted [a patch](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/11180) to the VTK upstream, so the root cause will be fixed by VTK team soon.

**Reply by georgy-moshkin** [reply]
> I'm really confused here, will going and doing my homework to understand the maths help me use the software and stop so much of this trial and error? I feel like I should be able to get something simple like a Patch antenna working without too much hassle.


Hey, let me show of what is possible:
https://www.youtube.com/watch?v=v8uxhbu36Ug
While I have not released any free examples yet, I have planned two informative openEMS tutorials related to patch antennas. I haven't encountered any strange port errors, probably because I rigorously implement meshing code and spent a lot of time to obtain results similar to few research papers. In some cases, I put 10 mesh lines on a feeding line between "1/3 rule" lines to get correct results, otherwise it is +-30 Ohm compared to correct results. This was really a breakthrough for me, because I've got the same results as in CST or HFSS graphs from one paper. Also, I use "algorithm 3" for recursive mesh smooth. During my experimentation, most problems was caused by smooth mesh routines together with detectedges routines, when there are two very closely spaced meshing lines forced algorithm to create unnecessarily dense meshing lines. Thus, in most cases I put all meshing lines manually, and then run smoothmesh with "algorithm 3" for nice smooth mesh generation. Also, for learning I think that coarse mesh lines are preferable, because it saves a lot of time. So you can try more ideas faster. For real antenna design I would use dense mesh.
